components/p3a/nitro_utils/attestation.cc

```diff
@@ -0,0 +1,264 @@
+        "is running."
+      data: "A random nonce, and attestation document"
+      destination: WEBSITE
+    }
+    policy {
+      cookies_allowed: NO
+      setting:
+        "This feature can be disabled via the P3A setting."
+      policy_exception_justification:
+        "Not implemented."
+    })");
+}
+
+bool VerifyNonce(const cbor::Value::MapValue& cose_map,
+                 const std::vector<uint8_t>& orig_nonce) {
+  cbor::Value::MapValue::const_iterator nonce_it =
+      cose_map.find(cbor::Value("nonce"));
+  if (nonce_it == cose_map.end() || !nonce_it->second.is_bytestring()) {
+    LOG(ERROR) << "Nitro verification: nonce is missing or is not bstr";
+    return false;
+  }
+
+  if (nonce_it->second.GetBytestring() != orig_nonce) {
+    LOG(ERROR) << "Nitro verification: nonce mismatch";
+    return false;
+  }
+  return true;
+}
+
+bool VerifyUserDataKey(scoped_refptr<net::X509Certificate> server_cert,
+                       const cbor::Value::MapValue& cose_map) {
+  cbor::Value::MapValue::const_iterator user_data_it =
+      cose_map.find(cbor::Value("user_data"));
+  if (user_data_it == cose_map.end() || !user_data_it->second.is_bytestring() ||
+      user_data_it->second.GetBytestring().size() != 32) {
+    LOG(ERROR) << "Nitro verification: user data is missing or is not bstr "
+               << "or is not 32 bytes";
+    return false;
+  }
+  net::SHA256HashValue server_cert_fp =
+      net::X509Certificate::CalculateFingerprint256(server_cert->cert_buffer());
+  if (memcmp(server_cert_fp.data, user_data_it->second.GetBytestring().data(),
+             32) != 0) {
+    LOG(ERROR)
+        << "Nitro verification: server cert fp does not match user data fp, "
+        << "user data = "
+        << base::HexEncode(user_data_it->second.GetBytestring())
+        << ", server cert fp = " << base::HexEncode(server_cert_fp.data);
+    return false;
+  }
+  return true;
+}
+
+std::unique_ptr<net::ParsedCertificateList> ParseCertificatesAndCheckRoot(
+    scoped_refptr<net::X509Certificate> server_cert,
+    const cbor::Value::MapValue& cose_map) {
+  cbor::Value::MapValue::const_iterator cert_it =
+      cose_map.find(cbor::Value("certificate"));
+  cbor::Value::MapValue::const_iterator cabundle_it =
+      cose_map.find(cbor::Value("cabundle"));
+  if (cert_it == cose_map.end() || cabundle_it == cose_map.end() ||
+      !cabundle_it->second.is_array()) {
+    LOG(ERROR) << "Nitro verification: certificate and/or cabundle are "
+               << "missing or not the right type";
+    return nullptr;
+  }
+
+  const cbor::Value::ArrayValue& cabundle_arr = cabundle_it->second.GetArray();
+
+  std::vector<cbor::Value> cert_vals;
+  cert_vals.push_back(cert_it->second.Clone());
+  std::transform(cabundle_arr.rbegin(), cabundle_arr.rend(),
+                 std::back_inserter(cert_vals),
+                 [](const cbor::Value& val) { return val.Clone(); });
+
+  std::unique_ptr<net::ParsedCertificateList> cert_chain =
```

Comment:
maybe just use it by value?

Reply:
i returned unique_ptr so we can return nullptr if an error occurs, i'll return optional instead